text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,558 @@
+`S3EventProperty` candidate also supports specifying values as a scalar or array (a developer experience improvement in
+languages where _type unions_ are available):
+
+```ts
+// @aws-cdk/aws-sam/lib/sam.generated.ts
+export namespace CfnFunction {
+  // ...
+  export interface EventSourceProperty {
+    // ...
+    readonly properties:
+      | CfnFunction.S3EventProperty
+      | CfnFunction.SNSEventProperty
+      | CfnFunction.SQSEventProperty
+      | CfnFunction.KinesisEventProperty
+      | CfnFunction.DynamoDBEventProperty
+      | CfnFunction.ApiEventProperty
+      | CfnFunction.ScheduleEventProperty
+      | CfnFunction.CloudWatchEventEventProperty
+      | CfnFunction.CloudWatchLogsEventProperty
+      | CfnFunction.IoTRuleEventProperty
+      | CfnFunction.AlexaSkillEventProperty
+      | CfnFunction.EventBridgeRuleEventProperty
+      | cdk.IResolvable;
+    // ...
+  }
+  // ...
+  export interface S3EventProperty {
+    // ...
+    readonly events: string[] | string | cdk.IResolvable;
+    // ...
+  }
+}
+```
+
+[cloudformation registry schema]:
+  https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-schema.html
+[json-schema-draft-07]: https://json-schema.org/draft-07/json-schema-release-notes.html
+[aws serverless application model]: https://github.com/awslabs/serverless-application-model
+[`aws::serverless::function` events]:
+  https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html#sam-function-events
+
+## Problems with the current implementation
+
+While Dynamic languages typically support _type unions_ without problems, statically typed languages such as **C#**,
+**Go** and **Java** lack native support for those types. The resulting developer experience degradations are detailed in
+this section, with a **Java** example based on the use-case examples listed above where possible.
+
+### Return values cannot be accurately typed
+
+When a method or property returns a _type union_, statically typed languages are simply unable to express that and must
+instead fall-back to the generic common super-type (e.g: `java.lang.Object` in **Java**):
+
+```java
+public class CfnFunction {
+  // ...
+  public interface EventSourceProperty {
+    // ...
+    public Object getProperties();
+    // ...
+  }
+  // ...
+}
+```
+
+This makes it more difficult for **Java** developers to write _correct_ and _safe_ code, since static type analysis
+cannot help identifying issues. Mistakes will typically result in a `ClassCastException` being thrown. This is also a
+degradation in the documentation experience, since the type signature of methods is no longer accurate.
+
+### Type-safe overloads cannot always be generated
+
+Since most statically typed languages support _overloads_ (although **Go** does not). This means we are often able to
+leverage overloads to provide type-safe entry points, for example in the **Java** fluent builders:
```

Comment:
Since most statically typed languages support _overloads_ (although **Go** does not), we are often able to
leverage overloads to provide type-safe entry points, for example in the **Java** fluent builders:

(Combine the sentences. Also is "fluent" correct?)

Reply:
Yeah [introduction to the Fluent Builder Pattern](https://dzone.com/articles/fluent-builder-pattern).